### PR TITLE
Desktop meta: strip ${SNAP} from .desktop icon

### PR DESCRIPTION
**Pull request by diddledan**
A lot of snaps are hacking the desktop file of their projects to insert a snap-relative path for their icon of the form `${SNAP}/path/to/icon.png`. The current .desktop extractor implementation does not understand this format so this commit strips the `${SNAP}` from the start of the path if it is present before continuing as before.

Signed-off-by: Daniel Llewellyn <daniel@bowlhat.net>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

**Comment by diddledan** [comment]
Test is added :-)

**Comment by diddledan** [comment]
seems the unit test I added is failing. I guess I still need to do some more work...

File: snapcraft/extractors/appstream.py [hunk 1]
```diff
@@ -194,6 +194,10 @@ def _get_icon_from_desktop_file(
         entry = DesktopEntry()
         entry.parse(os.path.join(workdir, path))
         icon = entry.getIcon()
+
+        if icon.startswith("${{SNAP}}"):  # matches '${SNAP}'.
+            icon = icon[7:]  # strip the '${SNAP}' prefix.
```
Comment (sergiusens) [inline snapcraft/extractors/appstream.py]: you also need to strip the leading slash for "joins" to work

Comment (diddledan) [reply snapcraft/extractors/appstream.py]: Let's see how the testsuite copes with that then. Changeset 4e56055 adds a trailing slash to the `${SNAP}`-stripping code.

File: snapcraft/extractors/appstream.py [hunk 2]
```diff
@@ -194,6 +194,11 @@ def _get_icon_from_desktop_file(
         entry = DesktopEntry()
         entry.parse(os.path.join(workdir, path))
         icon = entry.getIcon()
+
+        if icon.startswith("${{SNAP}}/"):  # matches '${SNAP}'.
```
Comment (sergiusens) [inline snapcraft/extractors/appstream.py]: I don't think this works, this is for formatting purposes.
If you want to be certain, then maybe do
```suggestion
        if icon.startswith(r"${SNAP}/"):
```

**Comment by diddledan** [comment]
looks like tests are still failing :-(

File: tests/unit/test_meta.py [hunk 3]
```diff
@@ -450,11 +454,20 @@ def test_create_meta_with_app_desktop_key(self):
...
+            Not(Contains(
+                "Icon {} specified in desktop file {} not found "
+                "in prime directory".format("${SNAP}/usr/share/app3.png", "app3.desktop")
+            )),
+        )
 
         snap_yaml = os.path.join("prime", "meta", "snap.yaml")
         self.assertThat(snap_yaml, Not(FileContains("desktop: app1.desktop")))
         self.assertThat(snap_yaml, Not(FileContains("desktop: app2.desktop")))
-        self.assertThat(snap_yaml, Not(FileContains("desktop: app3.desktop")))
```
Comment (sergiusens) [inline tests/unit/test_meta.py]: you still need to verify that the desktop key did not make it to snap.yaml, right?

File: snapcraft/extractors/appstream.py [hunk 4]
```diff
@@ -194,6 +194,7 @@ def _get_icon_from_desktop_file(
         entry = DesktopEntry()
         entry.parse(os.path.join(workdir, path))
         icon = entry.getIcon()
+
```
Comment (sergiusens) [inline snapcraft/extractors/appstream.py]: this empty line should probably be removed

File: tests/unit/test_meta.py (hunk 3 above)
Comment (diddledan) [reply tests/unit/test_meta.py]: oops, I didn't mean to remove that

**Comment by cevap** [comment]
Where did you test this one, it seems it broke several packages, fix delayed over a month ... . What do you mean by "test" or "tested"

**Pull request merged**